### PR TITLE
Paint Submarine Ride

### PR DESCRIPTION
**Pull request by marijnvdwerf**
image

### CODE REVIEW

File: src/ride/track_data.c [hunk 1]
```diff
@@ -5545,7 +5545,7 @@ const uint32 RideTypeTrackPaintFunctionsOld[91] = {
 	0x008AF764,						// RIDE_TYPE_AIR_POWERED_VERTICAL_COASTER
 	0x00890940,						// RIDE_TYPE_INVERTED_HAIRPIN_COASTER
 	0x00898384,						// RIDE_TYPE_MAGIC_CARPET
-	0x008995D4,						// RIDE_TYPE_SUBMARINE_RIDE
+	0/*x008995D4*/,						// RIDE_TYPE_SUBMARINE_RIDE
```
Comment (Gymnasiast) [inline src/ride/track_data.c]: You left this commented out. It should be removed when you're done.

Comment (marijnvdwerf) [inline src/ride/track_data.c]: Helper functions will need to be moved as well. Just wanted to get this out.

**Comment by Gymnasiast** [comment]
Is this a WIP?

**Comment by marijnvdwerf** [comment]
Well. All tracks are finished, but this probably needs some refactoring. Though that can be done in the mini helicopter PR as well. So kinda unsure myself.

**Comment by Gymnasiast** [comment]
Let's do the refactoring in the Mini Helicopters PR then, if you do it now you might do it again for the Mini Helicopters.

File: src/ride/track_data.c (hunk 1 above)
Comment (IntelOrca) [inline src/ride/track_data.c]: @marijnvdwerf are you going to remove this?

File: src/ride/water/submarine_ride.c [hunk 2]
```diff
@@ -64,3 +65,349 @@ void vehicle_visual_submarine(int x, int imageDirection, int y, int z, rct_vehic
...
+
+		hasFence = track_paint_util_has_fence(EDGE_NW, position, mapElement, ride, get_current_rotation());
+		imageId = (hasFence ? SPR_STATION_PIER_EDGE_NW_FENCED : SPR_STATION_PIER_EDGE_NW) | RCT2_GLOBAL(0x00F4419C, uint32);
+		sub_98197C(imageId, 0, 0, 32, 6, 1, height, 0, 2, height, get_current_rotation());
+		track_paint_util_draw_station_covers(EDGE_NW, hasFence, entranceStyle, direction, height);
+
+		imageId = SPR_STATION_PIER_EDGE_SE | RCT2_GLOBAL(0x00F4419C, uint32);
+		sub_98196C(imageId, 0, 24, 32, 8, 1, height, get_current_rotation());
+
+		if (RCT2_GLOBAL(0x0141E9DB, uint8) & 3) {
```
Comment (duncanspumpkin) [inline src/ride/water/submarine_ride.c]: I don't think that this is how the station pieces should be drawn. This doesn't make a difference for water rides but I feel we should stick to the same pattern. In my opinion when underground we should draw just like above ground but without the station covers. What happens in vanilla is that the station is drawn missing the front fence and drawing half of the station cover. This is clearly not quite right.

File: src/ride/track_paint.c [hunk 3]
```diff
@@ -185,6 +196,73 @@ void track_paint_util_paint_fences(uint8 edges, rct_xy16 position, rct_map_eleme
...
+	if ((!(position.x & (1 << 5))) && (!(position.y & (1 << 5))))
+		return true;
+
+	return false;
+}
+
+bool track_paint_util_draw_station_covers(enum edge edge, bool hasFence, const rct_ride_entrance_definition * entranceStyle, uint8 direction, uint16 height)
+{
+	uint32 imageId;
+	uint32 baseImageId = entranceStyle->flags;
```
Comment (duncanspumpkin) [inline src/ride/track_paint.c]: entranceStyle->flags should really be renamed to `baseImageId`.

File: src/ride/track_paint.c [hunk 4]
```diff
@@ -185,6 +196,73 @@ void track_paint_util_paint_fences(uint8 edges, rct_xy16 position, rct_map_eleme
...
+
+	if ((!(position.x & (1 << 5))) && (!(position.y & (1 << 5))))
+		return true;
+
+	return false;
+}
+
+bool track_paint_util_draw_station_covers(enum edge edge, bool hasFence, const rct_ride_entrance_definition * entranceStyle, uint8 direction, uint16 height)
+{
+	uint32 imageId;
```
Comment (duncanspumpkin) [inline src/ride/track_paint.c]: Add underground check to this code ```if (!(RCT2_GLOBAL(0x0141E9DB, uint8) & 3))return;```

**Pull request merged**